// @prettier/plugin-pug/dist/options/converge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convergeOptions = void 0;
function convergeOptions(options) {
    var _a, _b, _c, _d, _e, _f;
    return {
        printWidth: options.printWidth,
        pugPrintWidth: options.pugPrintWidth !== -1 ? options.pugPrintWidth : options.printWidth,
        singleQuote: options.singleQuote,
        pugSingleQuote: (_a = options.pugSingleQuote) !== null && _a !== void 0 ? _a : options.singleQuote,
        tabWidth: options.tabWidth,
        pugTabWidth: options.pugTabWidth !== -1 ? options.pugTabWidth : options.tabWidth,
        useTabs: options.useTabs,
        pugUseTabs: (_b = options.pugUseTabs) !== null && _b !== void 0 ? _b : options.useTabs,
        bracketSpacing: options.bracketSpacing,
        pugBracketSpacing: (_c = options.pugBracketSpacing) !== null && _c !== void 0 ? _c : options.bracketSpacing,
        arrowParens: options.arrowParens,
        pugArrowParens: (_d = options.pugArrowParens) !== null && _d !== void 0 ? _d : options.arrowParens,
        semi: options.semi,
        pugSemi: (_e = options.pugSemi) !== null && _e !== void 0 ? _e : options.semi,
        bracketSameLine: options.bracketSameLine,
        pugBracketSameLine: (_f = options.pugBracketSameLine) !== null && _f !== void 0 ? _f : options.bracketSameLine,
        pugAttributeSeparator: options.pugAttributeSeparator,
        pugCommentPreserveSpaces: options.pugCommentPreserveSpaces,
        pugSortAttributes: options.pugSortAttributes,
        pugSortAttributesBeginning: options.pugSortAttributesBeginning,
        pugSortAttributesEnd: options.pugSortAttributesEnd,
        pugWrapAttributesThreshold: options.pugWrapAttributesThreshold,
        pugWrapAttributesPattern: options.pugWrapAttributesPattern,
        pugClassNotation: options.pugClassNotation,
        pugIdNotation: options.pugIdNotation,
        pugClassLocation: options.pugClassLocation,
        pugEmptyAttributes: options.pugEmptyAttributes,
        pugEmptyAttributesForceQuotes: options.pugEmptyAttributesForceQuotes,
        pugSingleFileComponentIndentation: options.pugSingleFileComponentIndentation &&
            options.__embeddedInHtml === true,
        pugFramework: options.pugFramework,
        pugExplicitDiv: options.pugExplicitDiv,
    };
}
exports.convergeOptions = convergeOptions;
